### PR TITLE
api: Fix version detection for non-main-branch images

### PR DESCRIPTION
**Pull request by eseliger**
I noticed that some instances were still using api-version 1, and went to dig in why that is. Turns out some edge-case in our versioning scheme wasn't honored (branch name prefix), so I fixed that.

Test plan: Added tests that fail before the code change.

### CODE REVIEW

File: lib/shared/src/sourcegraph-api/siteVersion.test.ts [hunk 1]
```diff
@@ -16,6 +16,10 @@ describe('inferCodyApiVersion', () => {
         expect(inferCodyApiVersion('5.5.0', false)).toBe(1)
         expect(inferCodyApiVersion('5.6.0', false)).toBe(1)
         expect(inferCodyApiVersion('5.7.0', false)).toBe(1)
+        // branch deployment (2024-09-11 is the cutoff date for 2->8)
+        expect(inferCodyApiVersion('5.11.x_313350_2024-09-10_5.11-63a41475e780', false)).toBe(1)
+        // main deployment (2024-09-11 is the cutoff date for 2->8)
+        expect(inferCodyApiVersion('315302_2024-09-10_5.11-9994f058e2af', false)).toBe(1)
```
Comment (eseliger) [inline lib/shared/src/sourcegraph-api/siteVersion.test.ts:22]: After https://github.com/sourcegraph/cody/pull/7365, this changed to 8 from `1`, which means that older instances running untagged versions would now be forced to use version 8, which they don't support

File: lib/shared/src/sourcegraph-api/siteVersion.test.ts [hunk 2]
```diff
@@ -25,8 +29,18 @@ describe('inferCodyApiVersion', () => {
         expect(inferCodyApiVersion('5.10.1', false)).toBe(DefaultMinimumAPIVersion)
     })
 
-    test('returns latestCodyClientConfig for dotcom', () => {
+    test('Latest API version', () => {
+        // dotcom
         expect(inferCodyApiVersion('314951_2025-03-07_6.1-abeeb1a5e10d', true)).toBe(8)
+
+        // TODO: Erik believes these should return 8, but they don't do that anymore
+        // after https://github.com/sourcegraph/cody/pull/7365.
```
Comment (eseliger) [inline lib/shared/src/sourcegraph-api/siteVersion.test.ts:37]: cc @abeatrix re the comment on https://github.com/sourcegraph/cody/pull/7365, is this the desired behavior?

File: lib/shared/src/sourcegraph-api/siteVersion.test.ts [hunk 3]
```diff
@@ -25,8 +29,18 @@ describe('inferCodyApiVersion', () => {
...
+        expect(inferCodyApiVersion('6.1.0', false)).toBe(5)
+        expect(inferCodyApiVersion('6.2.0', false)).toBe(5)
+        // branch deployment
+        expect(inferCodyApiVersion('6.1.x_313350_2025-02-25_6.1-63a41475e780', false)).toBe(8)
+        // main deployment
+        expect(inferCodyApiVersion('315302_2025-03-10_6.1-9994f058e2af', false)).toBe(8)
```
Comment (eseliger) [inline lib/shared/src/sourcegraph-api/siteVersion.test.ts:43]: both are now correctly inferred to support API 8

File: lib/shared/src/sourcegraph-api/siteVersion.ts [hunk 4]
```diff
@@ -179,16 +179,17 @@ export function inferCodyApiVersion(version: string, isDotCom: boolean): CodyApi
...
-    // release less frequently.
+    // like "2024-09-11_5.7-4992e874aee2" or "6.1.x_313350_2025-02-25_6.1-63a41475e780",
+    // which does not parse as SemVer.  We make a best effort go parse the date from
+    // the version identifier allowing us to selectively enable new API versions on
+    // instances like SG02 (that deploy frequently) without crashing on other Cloud
+    // deployments that release less frequently.
     if (parsedVersion === null) {
         const date = parseDateFromPreReleaseVersion(version)
-        if (date && date < new Date('2024-09-11')) {
+        if (date && date >= new Date('2024-09-11')) {
```
Comment (eseliger) [inline lib/shared/src/sourcegraph-api/siteVersion.ts:189]: I believe this was inverted in https://github.com/sourcegraph/cody/pull/7365

File: lib/shared/src/sourcegraph-api/siteVersion.ts [hunk 5]
```diff
@@ -179,16 +179,17 @@ export function inferCodyApiVersion(version: string, isDotCom: boolean): CodyApi
...
             return 8
         }
+        return 1
```
Comment (eseliger) [inline lib/shared/src/sourcegraph-api/siteVersion.ts:192]: I added this back so that very old instances still use V1 and not V5 - lmk if this is wrong.

Comment (abeatrix) [reply lib/shared/src/sourcegraph-api/siteVersion.ts:192]: @bobheadxi is that ok or do we still want to set those with v5?

Comment (bobheadxi) [reply lib/shared/src/sourcegraph-api/siteVersion.ts:192]: I just know we need to make sure newer instances, dev instances, CD instances, etc need V5+ since that's when usage was added. I'm not sure what our policies for support or versioning is

File: lib/shared/src/sourcegraph-api/siteVersion.test.ts (hunk 1 above)
Comment (abeatrix) [reply lib/shared/src/sourcegraph-api/siteVersion.test.ts:22]: this only happens if setLatestCodyAPIVersion wasn't triggered though so that's the desired behavior right?

File: lib/shared/src/sourcegraph-api/siteVersion.ts [hunk 6]
```diff
@@ -179,16 +179,17 @@ export function inferCodyApiVersion(version: string, isDotCom: boolean): CodyApi
...
     }
 
     console.log(version, parsedVersion, 'inferCodyApiVersion')
```
Comment (abeatrix) [inline lib/shared/src/sourcegraph-api/siteVersion.ts:195]:
```suggestion
```

File: lib/shared/src/sourcegraph-api/siteVersion.test.ts (hunk 2 above)
Comment (abeatrix) [reply lib/shared/src/sourcegraph-api/siteVersion.test.ts:37]: Yea cause setLatestCodyAPIVersion should be triggered to set latestCodyClientConfig. if that wasn't triggered, then i think it's fine to fall back to the min version which is 5?

File: lib/shared/src/sourcegraph-api/siteVersion.ts (hunk 4 above)
Comment (abeatrix) [reply lib/shared/src/sourcegraph-api/siteVersion.ts:189]: should we return 5 instead of 8 here since 8 is not available until earlier last month

**Review by abeatrix — APPROVED** [review]
thank you!

**Comment by dominiccooney** [comment]
Backport fails to cleanly apply because of sourcegraph/cody#7365, I am going to assume that is also desired.
